compiler/rustc_metadata/src/native_libs.rs

```diff
@@ -396,4 +384,48 @@ impl Collector<'tcx> {
             }
         }
     }
+
+    fn i686_arg_list_size(&self, item: &hir::ForeignItemRef<'_>) -> usize {
+        let argument_types: &List<Ty<'_>> = self.tcx.erase_late_bound_regions(
+            self.tcx
+                .type_of(item.id.def_id)
+                .fn_sig(self.tcx)
+                .inputs()
+                .map_bound(|slice| self.tcx.mk_type_list(slice.iter())),
+        );
+
+        // stdcall on i686 never allocates less than 32 bits for an argument.
+        let min_arg_size: usize = 4;
+
+        argument_types.iter().fold(0, |accum: usize, ty: Ty<'_>| {
+            let layout = self
+                .tcx
+                .layout_of(ParamEnvAnd { param_env: ParamEnv::empty(), value: ty })
+                .expect("layout")
+                .layout;
+            accum + std::cmp::max(min_arg_size, layout.size.bytes_usize())
+        })
+    }
+
+    fn build_dll_import(&self, abi: Abi, item: &hir::ForeignItemRef<'_>) -> DllImport {
+        let calling_convention = if self.tcx.sess.target.arch == "x86" {
+            match abi {
+                Abi::C { .. } | Abi::Cdecl => DllCallingConvention::C,
+                Abi::Stdcall { .. } | Abi::System { .. } => {
+                    DllCallingConvention::Stdcall(self.i686_arg_list_size(item))
+                }
+                Abi::Fastcall => DllCallingConvention::Fastcall(self.i686_arg_list_size(item)),
+                // Vectorcall is intentionally not supported at this time.
+                _ => {
+                    self.tcx.sess.span_fatal(
+                        item.span,
+                        r#"ABI not supported by `#[link(kind = "raw-dylib")]` on i686"#,
+                    );
+                }
+            }
+        } else {
+            DllCallingConvention::C
```

Comment:
This logic is going to normalize arbitrary irrelevant ABIs (e.g. `extern "msp430-interrupt"`) to `extern "C"`.

Also see some discussion on accepting ABIs on various targets in https://github.com/rust-lang/rust/pull/86231.

Reply:
I read through the comments on the linked PR, and I'm a bit confused about the desired change here.  Should I signal an error on other cpu architectures for anything other than C, cdecl, stdcall, fastcall, and maybe vectorcall?  Or just signal an error for anything except C?  (MSVC ignores the other calling conventions on those platforms, but it's not clear that we should follow suit.)

It's probably worth pointing out that this function should only ever run on windows-msvc targets; I'll add a comment (or, better, an assertion) to that effect.

Reply:
Also, I notice that rustc supports an i586-pc-windows-msvc target.  Should we support stdcall & fastcall there also?

Reply:
>Also, I notice that rustc supports an i586-pc-windows-msvc target. Should we support stdcall & fastcall there also?

It should be automatically supported, `i586` is basically the same target as `i686` just with a different base CPU (controlling the set of available CPU features), see e.g. the target definition in `compiler\rustc_target\src\spec\i586_pc_windows_msvc.rs`.

Reply:
>I read through the comments on the linked PR, and I'm a bit confused about the desired change here.

I'm not entirely sure yet.
I suspect that MSVC supported the "legacy" conventions on x64 for compatibility reasons, to ease portability of huge amount of pre-existing code.
We don't have such a problem, so for us it would be better to reject all the "legacy" conventions on x64.

Reply:
Let's accept only `Abi::C` and `Abi::Win64` here for a start, and produce the "ABI not supported" error for anything else.

Reply:
(Note that vectorcall is a thing on x64, and it's different from the default C/win64, and I have no idea how it's mangled there, but this PR doesn't support vectorcall, so it's not a problem right now.)

Reply:
Re vectorcall: interesting; I wasn't aware that was a thing on 64-bit.  Good to know!

For context, I decided not to support vectorcall on 32-bit because there appear to be deeper problems with it than just getting the name mangling correct.  Making the obvious modification here to allow vectorcall and running the PR's run-make test (with, effectively, `s/stdcall/vectorcall/g;`) didn't work, but in an interesting way: it all links and runs, but the output indicates that the called function isn't seeing the same parameter values that the calling function provides.  So my hunch is that there are other issues with vectorcall in rustc that are outside the scope of this PR.  My primary motivation with the raw-dylib work is facilitating interop with the Windows API, and effectively all of those functions are stdcall on x86-32, so it just didn't seem worth it to dig into the vectorcall issues further.

Reply:
Also added `Abi::System` for x86_64, as that appears to be a synonym for `Abi::C` (at least on x86_64) that devs will expect to be able to use.